src/System.Net.NameResolution/src/System/Net/NameResolutionPal.Windows.cs

```diff
@@ -385,10 +383,299 @@ public static void EnsureSocketsAreInitialized()
+                {
+                    HostName = canonicalName,
+                    Aliases = Array.Empty<string>(),
+                    AddressList = addresses.ToArray()
+                });
+            }
+            finally
+            {
+                state.Dispose();
+            }
+        }
+
+        private static unsafe IPAddress CreateIPv4Address(byte* socketAddress, int addressLength)
+        {
+            const int IPv4SocketAddressSize = 16;
+
+            if (addressLength != IPv4SocketAddressSize)
+                return null;
+
+            long address = (long)(
+                               (socketAddress[4] & 0x000000FF) |
+                               (socketAddress[5] << 8 & 0x0000FF00) |
+                               (socketAddress[6] << 16 & 0x00FF0000) |
+                               (socketAddress[7] << 24)
+                           ) & 0x00000000FFFFFFFF;
+
+            return new IPAddress(address);
+        }
+
+        private static unsafe IPAddress CreateIPv6Address(byte* socketAddress, int addressLength)
+        {
+            const int IPv6SocketAddressSize = 28;
+            const int IPv6AddressBytes = 16;
+
+            if (addressLength != IPv6SocketAddressSize)
+                return null;
+
+            byte[] address = new byte[IPv6AddressBytes];
+            for (int i = 0; i < address.Length; i++)
+            {
+                address[i] = socketAddress[i + 8];
+            }
+
+            long scope = (long)((socketAddress[27] << 24) +
+                                (socketAddress[26] << 16) +
+                                (socketAddress[25] << 8) +
+                                (socketAddress[24]));
+
+            return new IPAddress(address, scope);
+        }
+
+        #region GetAddrInfoAsync Helper Classes
+
+        private sealed unsafe class GetAddrInfoExState : CriticalFinalizerObject, IDisposable
+        {
+            private IntPtr m_context;
+            private int m_callbackStartedOrCanceled;
+            private DnsResolveAsyncResult m_asyncResult;
+
+            public string HostAddress { get; }
+
+            public GetAddrInfoExState(string hostAddress, DnsResolveAsyncResult asyncResult)
+            {
+                HostAddress = hostAddress;
+                m_asyncResult = asyncResult;
+            }
+
+            [ReliabilityContract(Consistency.WillNotCorruptState, Cer.Success)]
+            public void SetQueryContext(IntPtr context)
+            {
+                m_context = context;
+            }
+
+            [ReliabilityContract(Consistency.WillNotCorruptState, Cer.Success)]
+            public bool SetCallbackStartedOrCanceled()
+            {
+                return Interlocked.Exchange(ref m_callbackStartedOrCanceled, 1) == 0;
+            }
+
+            public void CompleteAsyncResult(object o)
+            {
+                Task.Run(() => m_asyncResult.InvokeCallback(o));
```

Comment:
Do we need to queue the callback?

Assuming yes, this will allocate a delegate and closure.  It can instead be written as:
```C#
Task.Factory.StartNew(s =>
{
    var t = (Tuple<DnsResolveAsyncResult, object>)s;
    t.Item1(t.Item2);
}, Tuple.Create(_asyncResult, o), CancellationToken.None, TaskCreationOptions.None, TaskScheduler.Default);
```
That will incur the overhead of allocating just the tuple rather than the overhead of allocating the closure for the state as well as the delegate to a method on that closure.